File: plugins/verify/hooks/scripts/security-check.py
#!/usr/bin/env python3
"""
Pre-commit security check hook.
Warns about edits that might contain secrets or security issues.
All edits are allowed but warnings are shown to Claude.

Uses JSON output with permissionDecision: "allow" to warn without blocking.
"""

import json
import logging
import os
import re
import sys

logger = logging.getLogger(__name__)

# Patterns that indicate potential secrets
SECRET_PATTERNS = [
    (r'(?i)(api[_-]?key|apikey)\s*[:=]\s*["\']?[a-zA-Z0-9_-]{20,}', "API key"),
    (r'(?i)(secret|password|passwd|pwd)\s*[:=]\s*["\'][^"\']+["\']', "Password/Secret"),
    (r"(?i)bearer\s+[a-zA-Z0-9_-]{20,}", "Bearer token"),
    (r"ghp_[a-zA-Z0-9]{36}", "GitHub Personal Access Token"),
    (r"github_pat_[a-zA-Z0-9]{22}_[a-zA-Z0-9]{59}", "GitHub PAT (fine-grained)"),
    (r"sk-[a-zA-Z0-9]{48}", "OpenAI API Key"),
    (r"sk-ant-[a-zA-Z0-9-]{90,}", "Anthropic API Key"),
    (r"-----BEGIN (?:RSA |EC |DSA )?PRIVATE KEY-----", "Private key"),
    (r"(?i)aws[_-]?access[_-]?key[_-]?id\s*[:=]\s*[A-Z0-9]{20}", "AWS Access Key"),
    (
        r"(?i)aws[_-]?secret[_-]?access[_-]?key\s*[:=]\s*[a-zA-Z0-9/+=]{40}",
        "AWS Secret Key",
    ),
]

# Files to always skip
SKIP_FILES = {
    ".env.example",
    ".env.template",
    ".env.sample",
    "package-lock.json",
    "yarn.lock",
    "pnpm-lock.yaml",
}


def is_safe_path(file_path: str) -> bool:
    """
    Verify file_path resolves to a location within the current working directory.
    Uses os.path.commonpath to prevent path traversal attacks.
    """
    try:
        project_root = os.getcwd()
        target_path = os.path.abspath(os.path.join(project_root, file_path))
        return os.path.commonpath([project_root, target_path]) == project_root
    except Exception:
        return False


def check_for_secrets(content: str, file_path: str) -> list[str]:
    """Check content for potential secrets. Returns list of detected issues."""
    issues = []

    # Skip certain files
    if os.path.basename(file_path) in SKIP_FILES:
        return issues

    # Skip test files checking for secret patterns
    if "test" in file_path.lower() or "spec" in file_path.lower():
        return issues

    for pattern_str, secret_type in SECRET_PATTERNS:
        PATTERN = re.compile(pattern_str)
        matches = PATTERN.findall(content)
        if matches:
            issues.append(f"Potential {secret_type} detected")

    return issues


def main():
    try:
        input_data = json.load(sys.stdin)
        tool_input = input_data.get("tool_input", {})

        # Get file path and content based on tool type
        file_path = tool_input.get("file_path", "")
        content = tool_input.get("content", "") or tool_input.get("new_string", "")

        # Input validation
        if not file_path or not content:
            sys.exit(0)

        # Security: Check for path traversal attempts
        if not is_safe_path(file_path):
            logger.debug("[security-check] Path traversal detected: %s", file_path)
            sys.exit(0)

        issues = check_for_secrets(content, file_path)

        if issues:
            issues_text = "\n".join(f"  - {issue}" for issue in issues)
            warning = (
                f"[security-check] WARNING: Potential security issue detected in {file_path}:\n"
                f"{issues_text}\n"
                f"[security-check]   Please verify this is not a real secret before committing.\n"
                f"[security-check]   If this is a false positive, review and adjust patterns in security-check.py"
            )
            output = {
                "continue": True,
                "systemMessage": warning,
                "hookSpecificOutput": {
                    "hookEventName": "PreToolUse",
                    "permissionDecision": "allow",
                    "permissionDecisionReason": warning,
                },
            }
            print(json.dumps(output))
            sys.exit(0)

    except json.JSONDecodeError:
        # Invalid JSON input, don't block
        sys.exit(0)
    except Exception as e:
        # Don't block on errors, but log for debugging
        logger.debug("[security-check] Unexpected error: %s", e, exc_info=True)
        sys.exit(0)


if __name__ == "__main__":
    main()

File: plugins/guard-ts/hooks/scripts/security-check.js
#!/usr/bin/env node
/**
 * Pre-commit security check hook.
 * Warns about edits that might contain secrets or security issues.
 */
const fs = require('fs');
const path = require('path');

const SECRET_PATTERNS = [
    { regex: /(api[_-]?key|apikey)\s*[:=]\s*["']?[a-zA-Z0-9_-]{20,}/i, type: "API key" },
    { regex: /(secret|password|passwd|pwd)\s*[:=]\s*["'][^"']+["']/i, type: "Password/Secret" },
    { regex: /bearer\s+[a-zA-Z0-9_-]{20,}/i, type: "Bearer token" },
    { regex: /ghp_[a-zA-Z0-9]{36}/, type: "GitHub Personal Access Token" },
    { regex: /github_pat_[a-zA-Z0-9]{22}_[a-zA-Z0-9]{59}/, type: "GitHub PAT (fine-grained)" },
    { regex: /sk-[a-zA-Z0-9]{48}/, type: "OpenAI API Key" },
    { regex: /sk-ant-[a-zA-Z0-9-]{90,}/, type: "Anthropic API Key" },
    { regex: /-----BEGIN (?:RSA |EC |DSA )?PRIVATE KEY-----/, type: "Private key" },
    { regex: /aws[_-]?access[_-]?key[_-]?id\s*[:=]\s*[A-Z0-9]{20}/i, type: "AWS Access Key" },
    { regex: /aws[_-]?secret[_-]?access[_-]?key\s*[:=]\s*[a-zA-Z0-9/+=]{40}/i, type: "AWS Secret Key" }
];

const SKIP_FILES = new Set([
    ".env.example", ".env.template", ".env.sample",
    "package-lock.json", "yarn.lock", "pnpm-lock.yaml"
]);

function checkForSecrets(content, filePath) {
    const issues = [];
    const basename = path.basename(filePath);

    if (SKIP_FILES.has(basename)) return issues;
    // Skip test files
    if (filePath.toLowerCase().includes('test') || filePath.toLowerCase().includes('spec')) return issues;

    for (const { regex, type } of SECRET_PATTERNS) {
        if (regex.test(content)) {
            issues.push(`Potential ${type} detected`);
        }
    }
    return issues;
}

async function main() {
    try {
        const inputData = JSON.parse(fs.readFileSync(0, 'utf-8'));
        const toolInput = inputData.tool_input || {};

        const filePath = toolInput.file_path || '';
        const content = toolInput.content || toolInput.new_string || '';

        if (!filePath || !content) process.exit(0);

        const projectRoot = process.cwd();
        const targetPath = path.resolve(projectRoot, filePath);
        if (!path.relative(projectRoot, targetPath).startsWith('..')) {
            // Safe path
        } else {
            // Unsafe path
            process.exit(0);
        }

        const issues = checkForSecrets(content, filePath);

        if (issues.length > 0) {
            const issuesText = issues.map(i => `  - ${i}`).join('\n');
            const warning = `[security-check] WARNING: Potential security issue detected in ${filePath}:\n` +
                `${issuesText}\n` +
                `[security-check]   Please verify this is not a real secret before committing.\n` +
                `[security-check]   If this is a false positive, review security-check.js`;

            console.log(JSON.stringify({
                continue: true,
                systemMessage: warning,
                hookSpecificOutput: {
                    hookEventName: "PreToolUse",
                    permissionDecision: "allow",
                    permissionDecisionReason: warning
                }
            }));
            process.exit(0);
        }

    } catch (e) {
        process.exit(0);
    }
}

main();
